daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+  type DataValueBigInt = DataValue[JBigInt, NonNullable with DataValuePrimitiveType]
+  type DataValueString = DataValue[JString, NonNullable with DataValuePrimitiveType]
+  type DataValueURI = DataValue[URI, NonNullable with DataValuePrimitiveType]
+  type DataValueFloat = DataValue[JFloat, NonNullable with DataValuePrimitiveType]
+  type DataValueByte = DataValue[JByte, NonNullable with DataValuePrimitiveType]
+  type DataValueInt = DataValue[JInt, NonNullable with DataValuePrimitiveType]
+  type DataValueShort = DataValue[JShort, NonNullable with DataValuePrimitiveType]
+  type DataValueDINode = DataValue[DINode, NonNullable with DataValuePrimitiveType]
+  type DataValueUseNilForDefault = DataValue[UseNilForDefaultObj, NonNullable]
+
+  import scala.language.implicitConversions
+
+  @inline implicit def toDataValue(v: JBigDecimal): DataValueBigDecimal = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLCalendar): DataValueCalendar = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLDateTime): DataValueDateTime = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLDate): DataValueDate = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLTime): DataValueTime = new DataValue(v)
+  @inline implicit def toDataValue(v: Array[Byte]): DataValueByteArray = new DataValue(v)
+  @inline implicit def toDataValue(v: JBoolean): DataValueBool = new DataValue(v)
+  @inline implicit def toDataValue(v: JNumber): DataValueNumber = new DataValue(v)
+  @inline implicit def toDataValue(v: JLong): DataValueLong = new DataValue(v)
+  @inline implicit def toDataValue(v: JDouble): DataValueDouble = new DataValue(v)
+  @inline implicit def toDataValue(v: JBigInt): DataValueBigInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JString): DataValueString = new DataValue(v)
+  @inline implicit def toDataValue(v: URI): DataValueURI = new DataValue(v)
+  @inline implicit def toDataValue(v: JFloat): DataValueFloat = new DataValue(v)
+  @inline implicit def toDataValue(v: JByte): DataValueByte = new DataValue(v)
+  @inline implicit def toDataValue(v: JInt): DataValueInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JShort): DataValueShort = new DataValue(v)
+  @inline implicit def toDataValue(v: DINode): DataValueDINode = new DataValue(v)
+
+  @inline implicit def toDataValue(v: Long): DataValueLong = new DataValue(v: JLong)
+  @inline implicit def toDataValue(v: Double): DataValueDouble = new DataValue(v: JDouble)
+  @inline implicit def toDataValue(v: Boolean): DataValueBool = new DataValue(v: JBoolean)
+  @inline implicit def toDataValue(v: Float): DataValueFloat = new DataValue(v: JFloat)
+  @inline implicit def toDataValue(v: Byte): DataValueByte = new DataValue(v: JByte)
+  @inline implicit def toDataValue(v: Int): DataValueInt = new DataValue(v: JInt)
+  @inline implicit def toDataValue(v: Short): DataValueShort = new DataValue(v: JShort)
+
```

Comment:
Am I right that you didn't provide DataValue.apply(v: Long) = new DataValue(v: Long) for each acceptable DataValue anyref type?

So I have to write DataValue.toDataValue(myLong) instead of DataValue(myLong) ? 

Or maybe with these implicits you can just have a DataValue.apply(dv: DataValue) = dv and you pass any of the accepted types and it auto coerces them to a DataValue ? 

Or maybe you just don't need this. The implicits are enough and you never need an explicit constructor?

I'm not suggesting a change here necessarily at all. I just fully expected object DataValue to have a flurry of apply methods.

Reply:
I just relied entirely on the implicits.

Given the implicit, `DataValue(x)` would be equivalent to `x:DataValueAny` (or whatever suitable specialization covers the set of types you wrote apply for).

I have not even found any use for the above approach, instead finding it much clearer to put the annotation of the variable that gets written to. Eg:

val x:DataValuePrimitive = bar match{
 case a => 1
 case b => "a"
}

Reply:
I just relied entirely on the implicits.

Given the implicit, `DataValue(x)` would be equivalent to `x:DataValueAny` (or whatever suitable specialization covers the set of types you wrote apply for).

I have not even found any use for the above approach, instead finding it much clearer to put the annotation of the variable that gets written to. Eg:

val x:DataValuePrimitive = bar match{
 case a => 1
 case b => "a"
}

The one exception is a cast from AnyRef, which does require DataValue.unsafeFromAnyRef. In this case, I think we do not want an apply method, to force the developer to make sure that this is what they want to do.